modules/frontend/querysharding.go

```diff
@@ -167,3 +154,41 @@ func DoRequests(ctx context.Context, downstream Handler, reqs []*http.Request) (
 
 	return resps, firstErr
 }
+
+func mergeResponses(rrs []RequestResponse) (*http.Response, error) {
+	var errCode = http.StatusOK
+	var errBody io.ReadCloser
+	var combinedTrace []byte
+	for _, rr := range rrs {
+		if rr.Response.StatusCode == http.StatusOK {
+			body, err := ioutil.ReadAll(rr.Response.Body)
+			if err != nil {
+				return nil, errors.Wrap(err, "error ready response body at query frontend")
+			}
+
+			if len(combinedTrace) == 0 {
+				combinedTrace = body
+			} else {
+				combinedTrace = util.CombineTraces(combinedTrace, body)
+			}
+		}
+		if rr.Response.StatusCode > errCode {
+			errCode = rr.Response.StatusCode
+			errBody = rr.Response.Body
+		}
+	}
+
+	if len(combinedTrace) == 0 {
```

Comment:
If one of the queries 500s, but another query returns some data we will return 200 OK along with whatever was returned from the one query.

I feel like we should 500 in that case.

Reply:
Agree, propagating errors clearly seems best, rather than masking over the error and returning partial trace.